core/src/container.cpp

```diff
@@ -197,43 +197,56 @@ void ContainerBasePrivate::copyState(Interface::iterator external, const Interfa
 
 	// create a clone of external state within target interface (child's starts() or ends())
 	auto internal = states_.insert(states_.end(), InterfaceState(*external));
-	target->add(*internal);
 	// and remember the mapping between them
 	internalToExternalMap().insert(std::make_pair(&*internal, &*external));
+	target->add(*internal);
 }
 
 void ContainerBasePrivate::liftSolution(const SolutionBasePtr& solution, const InterfaceState* internal_from,
-                                        const InterfaceState* internal_to) {
-	computeCost(*internal_from, *internal_to, *solution);
-
-	// map internal to external states
-	auto find_or_create_external = [this](const InterfaceState* internal, bool& created) -> InterfaceState* {
-		auto it = internalToExternalMap().find(internal);
-		if (it != internalToExternalMap().end())
-			return const_cast<InterfaceState*>(it->second);
-
-		InterfaceState* external = &*states_.insert(states_.end(), InterfaceState(*internal));
-		internalToExternalMap().insert(std::make_pair(internal, external));
-		created = true;
+                                        const InterfaceState* internal_to, InterfaceState* new_from,
+                                        InterfaceState* new_to) {
+	const bool create_from{ requiredInterface().testFlag(WRITES_PREV_END) };
+	const bool create_to{ requiredInterface().testFlag(WRITES_NEXT_START) };
+
+	// external states, nullptr if they don't exist yet
+	const InterfaceState* external_from{ create_from ? new_from : internalToExternalMap().at(internal_from) };
+	const InterfaceState* external_to{ create_to ? new_to : internalToExternalMap().at(internal_to) };
```

Comment:
I think, here you assume that internalToExternalMap() is only used to link external and internal states for _reading_ interfaces.
If the interface is writing, i.e. `create_*` is true, but `new_*` is a `nullptr` (like in [`liftSolution()` calls from `SerialContainer`](https://github.com/v4hn/moveit_task_constructor/blob/29c27fcf71fb3129f87a450e4b8020dc6824dd93/core/src/container.cpp)) the resulting `external_*` state will be a `nullptr` as well. Further below, you will pass these null pointers to `storeSolution`, [which passes them on to `ContainerBasePrivate::onNewFailure`](https://github.com/v4hn/moveit_task_constructor/blob/bc03656ead4a79bc4ecce3a7a707b74d2cfbc130/core/src/stage.cpp).